snapcraft/internal/meta/_snap_packaging.py

```diff
@@ -513,14 +525,29 @@ def _generate_desktop_file(self, name, app):
             desktop_file.parse_and_reformat()
             desktop_file.write(gui_dir=os.path.join(self.meta_dir, 'gui'))
 
-    def _render_socket_modes(self, apps):
+    def _render_socket_modes(self, apps: Dict[str, Any]) -> None:
         for app in apps.values():
             sockets = app.get('sockets', {})
             for socket in sockets.values():
                 mode = socket.get('socket-mode')
                 if mode is not None:
                     socket['socket-mode'] = OctInt(mode)
-        return apps
+
+    def _apply_pass_through(self, location: str, this: Dict[str, Any],
+                            pass_through: Dict[str, Any],
+                            original: Dict[str, Any]) -> None:
+        # Look for properties defined in pass-through and add them to the
+        # given dictionary. Any value already in the original dictionary must
+        # not be specified in pass-through at the same time.
+        for key in pass_through:
+            if key in original:
+                raise meta_errors.AmbiguousPassThroughKeyError(key)
+            this[key] = pass_through[key]
+        if pass_through:
+            logger.warn('Pass-through is being used to add experimental '
+                        'properties to {!r} that have not been '
+                        'validated. The snap cannot be released to the store.'
+                        .format(location))
```

Comment:
This prints twice if one uses this feature in both the root and in apps. I think they'll probably know where they're using it, so I suggest only printing once if we see this keyword (no matter how many times we see it) and not specifying where it was.

Reply:
The point of the message is to very clearly point out that the feature is being used, right? Whether or not they remember using it, they should be reminded that this is something that will need to be removed in the future. So I think we don't want it to be less obtrusive.

Reply:
I think the real goal is to let them know that, because this keyword is being used, the store will flag it. Why do we need to say that multiple times?